drivers/scheduler/anthos/anthos.go

```diff
@@ -529,6 +543,126 @@ func (anth *anthos) unsetUserNameAndKey() error {
+	poolMap, err := getNodesSortByAge()
+	if err != nil {
+		return err
+	}
+	for _, sortedNodes := range poolMap {
+		startTime := initNodeUpgradeTime
+		for _, node := range sortedNodes {
+			diff := node.CreationTimestamp.Sub(startTime)
+			if diff > errorTimeDuration {
+				return fmt.Errorf("[%s] node upgrade took: [%v] minutes which is longer than the expected timeout value: [%v]",
+					node.Name, diff, errorTimeDuration)
+			}
+			log.Infof("[%s] node took: [%v] time to upgrade the node", node.Name, diff)
+			startTime = node.CreationTimestamp.Time
+		}
+	}
+	return nil
+}
+
+// getUserClusterName return Anthos user cluster name
+func (anth *anthos) getUserClusterName() (string, error) {
+	log.Info("Retrieving user cluster name")
+	var userCluster string
+	// Listing user cluster to get user cluster name
+	cmd := fmt.Sprintf("%s --kubeconfig %s clusters", listUserClustersCmd, adminKubeconfPath)
+	out, err := anth.execOnAdminWSNode(cmd)
+	if err != nil {
+		return "", fmt.Errorf("listing user clusters is failing: [%s]. Err: (%v)", out, err)
+	}
+	userClusters := strings.Split(out, "\n")
+	for _, cluster := range userClusters {
+		if strings.Contains(cluster, "NAME") {
+			continue
+		}
+		clusterInfo := strings.Fields(cluster)
+		userCluster = clusterInfo[1]
+		break
+	}
+	if userCluster == "" {
+		return "", fmt.Errorf("failed to find user cluster name")
+	}
+	log.Infof("Successfully retrieved user cluster name: [%s]", userCluster)
+	return userCluster, nil
+}
+
+// getStartTimeForNodePoolUpgrade return start time when node pool upgrade started
+func (anth *anthos) getStartTimeForNodePoolUpgrade(userClusterName string) (time.Time, error) {
+	log.Info("Getting start time for node pool upgrade")
+	var timeMatchReg = regexp.MustCompile(`\d+-\d+-\d+T\d+:\d+:\d+Z`)
+	var nodeUpgradeStartedReg = regexp.MustCompile(`gke-on-prem-last-upgrade-start-time: .+`)
+	const layout = "2006-01-02T15:04:05Z"
+	// Describe user cluster command provide last upgrade start time
+	cmd := fmt.Sprintf("%s --kubeconfig %s --cluster %s",
+		userClusterDescribeCmd, adminKubeconfPath, userClusterName)
+	log.Debugf("Executing command: %s", cmd)
+	out, err := anth.execOnAdminWSNode(cmd)
+	if err != nil {
+		return time.Time{}, fmt.Errorf("describing user cluster is failing: [%s]. Err: (%v)", out, err)
+	}
+	matches := nodeUpgradeStartedReg.FindAllString(out, -1)
+	if matches == nil {
+		return time.Time{}, fmt.Errorf("failed to match last-upgrade-start-time in describe user cluster command")
+	}
+	matchTimeInMinute := timeMatchReg.FindAllString(matches[0], -1)
+	if matchTimeInMinute == nil {
+		return time.Time{}, fmt.Errorf("failed to parse start time")
+	}
+	startTime, err := time.Parse(layout, matchTimeInMinute[0])
+	log.Debugf("Successfully retrieved startTime: %s", startTime)
+	return startTime, nil
+}
+
+// updateFileOwnership change file ownership to ubuntu users
+func (anth *anthos) updateFileOwnership() error {
```

Comment:
You can have this function take the directory name as param, that way you can reuse this function to update ownership on any directory.

Reply:
Incorporated !!